docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,244 @@
+
+[![Diagram](https://docs.google.com/drawings/d/e/2PACX-1vSZjddrWBkf2ynczeIMSzGo-yqe7Cg9X-TG_LjyWAP_ToyZolMEG_UnbNlvoSetRhDLmlfABd_rJ4W9/pub?w=960&h=720)](https://docs.google.com/drawings/d/1leRN6GUvU98Ur2TowMefk4ycTFNBRi9t1-dGtOPmHg0/edit?usp=sharing)
+
+## Risks
+
+* What if someone will set too low `syncDelay`? And upload itself will take more than `syncDelay` time. Block will be assumed as `ToDelete` state and will be removed. Thanks of `deleteDelay` it might have still some time to recover.How to ensure that will not happen? 
+  * Minimum syncDelay? 
+  * Grab time for syncDelay from object meta instead? Not from ULID? That is quite hard.
+  * Add safeguard on upload path? `upload timeout` has to smaller than `syncDelay`.
+
+* What if one block is malformed. Readers cannot handle it and crashes. How repair procedure will work? We can have repair process that can download block locally, rewrite it and fix it and upload. Problem is that it will take `syncDelay` time to
+appear in system. Since we are blocked, we need to make the block available immediately, ignoring eventual consistency limitations.
+Potential solutions:
+  * Just delete problematic block and live with 15 minute of delay for not having some portion of metrics? Will it not cause any downsampling issues?
+  * Use `ignore_delay` option and avoid Fresh state as possible. Eventual consistency issue may hit us.
+ 
+* Do we allow repair procedure when compactor is running? This is quite unsafe as compaction/downsamlping operation might be in progress 
+so repair procedure might be avoided. Turning off compactor might be tricky as well in potential HA mode. Maybe we want to introduce 
+"locking" some external labels from compactor operations. If yes, how to do it in eventual consistency system?
+
+## Action Items:
+
+*Ensure our block operations are following those rules (e.g create methods that ensures this).
+*Tests edge cases.
+
+## Alternatives:
+
+### Require strong consistent object storage.
+
+As mentioned in #Motivation section this would block community adoption as only a few of object storages has strong or even
+ clearly stated consistency guarantees.
+
+### Full immutability 
+
+By not removing any object, we may simplify problem. However we need hard deletions are:
+- enormously reducing storage size (e.g retention is exactly for this)
+- enormously reducing number of blocks in storage, improving syncing time.
+
+### Special integrity JSON file
+
+We may want to add some file to commit and mark integrity of the block. That file may hold some information like
+files we expect to have in the directory and maybe some hashes of those. 
+
+This was rejected for we can reuse existing meta.json as "integrity" marker (e.g if after 15m we don't have meta file, it is partially uploaded and should be removed).
+
+Additional information like expected number of files can be added later if needed, however with rule no 3. we should not need that.
```

Comment:
I think rule3 (meta file as first to be upload) may have to be refactored to talk about som content of that file that constitutes a checksum.

Reply:
Responded above, I think checksum is not helpful here